port_drayage_plugin/src/port_drayage_worker.cpp

```diff
@@ -182,20 +212,44 @@ namespace port_drayage_plugin
         }
     }
 
+    void PortDrayageWorker::update_cargo_information_after_action_completion(const PortDrayageMobilityOperationMsg& previous_port_drayage_msg) {
+        // If the previously received message was for 'Pickup' or 'Dropoff', update this object's _has_cargo flag and _cargo_id members accordingly
+        // Note: This assumes the previous 'Pickup' or 'Dropoff' action was successful
+        if (previous_port_drayage_msg.operation == PORT_DRAYAGE_PICKUP_OPERATION_ID) {
+            _has_cargo = true;
+
+            if (previous_port_drayage_msg.cargo_id) {
+                _cargo_id = *previous_port_drayage_msg.cargo_id;
+                ROS_DEBUG_STREAM("CMV completed pickup action. CMV is now carrying cargo " << _cargo_id);
+            }
+            else {
+                ROS_DEBUG_STREAM("CMV has completed pickup, but there is no Cargo ID associated with the picked up cargo.");
+                throw std::invalid_argument("CMV has completed pickup, but there is no Cargo ID associated with the picked up cargo.");
+            }
+        }
+        else if (previous_port_drayage_msg.operation == PORT_DRAYAGE_DROPOFF_OPERATION_ID) {
+            ROS_DEBUG_STREAM("CMV completed dropoff action. CMV is no longer carrying cargo " << _cargo_id);
+            _has_cargo = false;
+            _cargo_id = ""; // Empty string is used when no cargo is being carried
+        }
+    }
+
     void PortDrayageWorker::mobility_operation_message_parser(std::string mobility_operation_strategy_params) {
         // Use Boost Property Tree to parse JSON-encoded strategy_params field in MobilityOperations message
         using boost::property_tree::ptree;
         ptree pt;
         std::istringstream mobility_operation_strategy_params_ss(mobility_operation_strategy_params);
         boost::property_tree::json_parser::read_json(mobility_operation_strategy_params_ss, pt);
 
-        _latest_mobility_operation_msg.has_cargo = pt.get<bool>("cargo");
+        // Parse 'operation' field and assign the PortDrayageEvent type for this message accordingly
         _latest_mobility_operation_msg.operation = pt.get<std::string>("operation");
         ROS_DEBUG_STREAM("operation: " << _latest_mobility_operation_msg.operation);
-        ROS_DEBUG_STREAM("cargo flag: " << _latest_mobility_operation_msg.has_cargo);
-
-        if(_latest_mobility_operation_msg.operation == "MOVING_TO_LOADING_AREA") {
+        if (_latest_mobility_operation_msg.operation == PORT_DRAYAGE_PICKUP_OPERATION_ID) {
             _latest_mobility_operation_msg.port_drayage_event_type = PortDrayageEvent::RECEIVED_NEW_DESTINATION;
+            if (_has_cargo) {
+                ROS_DEBUG_STREAM("Received 'PICKUP' operation, but CMV is already carrying cargo.");
+                throw std::invalid_argument("Received 'PICKUP' operation, but CMV is already carrying cargo.");
```

Comment:
Are we guaranteed that CARMA streets will only ever broadcast this message once? I think otherwise we're liable to throw an uncaught exception here. Is this filtered by vehicle mobility ID as well? Or would we throw this exception if another vehicle was executing the pickup operation while we had cargo?

Reply:
Yes, CARMA Streets will only ever broadcast this message once, and PortDrayageWorker filters these messages by cmv_id, so a PICKUP message broadcasted to another vehicle would not cause an exception to be thrown here. Additionally, a message will not be processed if its strategy_params match the previously processed message's strategy_params.